frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -509,94 +539,120 @@ static void checkForParenlessMethodFieldRedefinition(Context* context,
-    for (auto formal : fn->formals()) {
-      formal->traverse(visitor);
+    // The parent signature must exist.
+    if (!parentSignature) return nullptr;
+
+    // All parents need to be concrete for the signature to be evaluated.
+    for (auto up = parentSignature; up; up = up->parentFn()) {
+      if (up->needsInstantiation()) return nullptr;
     }
-    // do not visit the return type or function body
+  }
 
-    // now, construct a TypedFnSignature from the result
-    std::vector<types::QualifiedType> formalTypes = visitor.getFormalTypes(fn);
-    bool needsInstantiation = anyFormalNeedsInstantiation(context, formalTypes,
-                                                          untypedSig,
-                                                          nullptr);
+  ResolutionResultByPostorderID r;
+  auto visitor = Resolver::createForInitialSignature(rc, fn, r);
 
-    // visit the where clause, unless it needs to be instantiated, in
-    // which case we will visit the where clause when that happens
-    TypedFnSignature::WhereClauseResult whereResult =
-      TypedFnSignature::WHERE_NONE;
-    if (auto whereClause = fn->whereClause()) {
-      if (needsInstantiation) {
-        whereResult = TypedFnSignature::WHERE_TBD;
-      } else {
-        whereClause->traverse(visitor);
-        whereResult = whereClauseResult(context, fn, r, needsInstantiation);
-      }
-    }
+  // visit the formals, but not the return type or body
+  for (auto formal : fn->formals()) formal->traverse(visitor);
 
-    checkForParenlessMethodFieldRedefinition(context, fn, visitor);
+  // Give up if we could not type outer variables present in the signature.
+  if (parentSignature && rc->isEmpty() && !visitor.outerVariables.isEmpty()) {
+    return nullptr;
+  }
 
-    result = TypedFnSignature::get(context,
-                                   untypedSig,
-                                   std::move(formalTypes),
-                                   whereResult,
-                                   needsInstantiation,
-                                   /* instantiatedFrom */ nullptr,
-                                   /* parentFn */ parentFnTyped,
-                                   /* formalsInstantiated */ Bitmap());
+  // now, construct a TypedFnSignature from the result
+  std::vector<types::QualifiedType> formalTypes = visitor.getFormalTypes(fn);
+  bool needsInstantiation = anyFormalNeedsInstantiation(context, formalTypes,
+                                                        untypedSig,
+                                                        nullptr);
+
+  // visit the where clause, unless it needs to be instantiated, in
+  // which case we will visit the where clause when that happens
+  auto whereResult = TypedFnSignature::WHERE_NONE;
+  if (auto whereClause = fn->whereClause()) {
+    if (needsInstantiation) {
+      // Visit the where clause for generic nested functions just to collect
+      // outer variables. TODO: Wrap in speculative block?
```

Comment:
Better to avoid a speculative block if we can.

This case seems quite tricky when combined with POI. I recommend you make a Future Work item about adding tests in this area.